### ISSUE TITLE
LSTM numerical inconsistency: float16 dtype produces 4.1e-04 relative L2 error  compared to float32 on CUDA (exceeds 1e-4 threshold)

### ISSUE THREAD

**Issue by zifan6699** [pytorch/pytorch#175845]
### 🐛 Describe the bug

When running a single LSTM layer with float16 dtype on CUDA,  the output differs from float32 by a relative L2 error of 4.1e-4. While some precision loss is expected when converting from float32 to float16, this error magnitude exceeds the 1e-4 threshold, which appears excessive. In deep models, such errors can accumulate and potentially affect model correctness and inference consistency.

```python
import torch
import torch.nn as nn
import os
import contextlib
import copy

torch.manual_seed(42)
torch.cuda.manual_seed_all(42)

@contextlib.contextmanager
def set_inductor_env_vars():
    original = os.environ.get("INDUCTOR_SHAPE_PADDING")
    try:
        os.environ["INDUCTOR_SHAPE_PADDING"] = "1"
        yield
    finally:
        if original is None:
            os.environ.pop("INDUCTOR_SHAPE_PADDING", None)
        else:
            os.environ["INDUCTOR_SHAPE_PADDING"] = original

def compare(t_a, t_b):
    a, b = t_a.detach().cpu().float(), t_b.detach().cpu().float()
    max_abs = torch.abs(a - b).max().item()
    l2_a, l2_b = torch.norm(a), torch.norm(b)
    rel_l2 = (torch.norm(a - b) / torch.maximum(l2_a, l2_b).clamp(min=1e-8)).item()
    return max_abs, rel_l2

def run_mode(lstm_layer, x_in, dev, dtype, opt):
    lstm = copy.deepcopy(lstm_layer).to(dev).to(dtype).eval()
    if opt != "eager":
        with set_inductor_env_vars():
            lstm = torch.compile(lstm)
    with torch.no_grad():
        return lstm(x_in.to(dev).to(dtype))[0]

def run():
    x_in = torch.randn(64, 10, 50)
    lstm_layer = nn.LSTM(input_size=50, hidden_size=50, num_layers=1, batch_first=True, bidirectional=False)
    
    print("Input: shape={}".format(x_in.shape))
    print("Input - min: {:.6f}, max: {:.6f}, mean: {:.6f}".format(x_in.min().item(), x_in.max().item(), x_in.mean().item()))
    print("Input - has NaN: {}, has Inf: {}".format(torch.isnan(x_in).any().item(), torch.isinf(x_in).any().item()))
    print("=" * 60)
    
    modes = [
        ("Mode A (cuda/f32/eager)", 'cuda', torch.float32, 'eager'),
        ("Switch Dtype (cuda/f16/eager)", 'cuda', torch.float16, 'eager')
    ]
    
    outputs = {}
    for label, dev, dtype, opt in modes:
        print(f"\n>>> Running {label} ...")
        outputs[label] = run_mode(lstm_layer, x_in, dev, dtype, opt)
        has_nan = torch.isnan(outputs[label]).any().item()
        has_inf = torch.isinf(outputs[label]).any().item()
        print(f"  ✓ {label} executed successfully")
        print(f"  Output - min: {outputs[label].min().item():.6f}, max: {outputs[label].max().item():.6f}, mean: {outputs[label].mean().item():.6f}")
        print(f"  Output - has NaN: {has_nan}, has Inf: {has_inf}")
    
    ma_name = modes[0][0]
    mb_name = modes[1][0]
    max_d, max_rd = compare(outputs[ma_name], outputs[mb_name])
    exceeds = "Yes" if max_rd > 1e-4 else "No"
    status = "DIFF > 1e-4" if max_rd > 1e-4 else "OK"
    
    print("\n" + "="*120)
    print(f"Comparison: {ma_name} vs {mb_name}")
    print("="*120)
    print(f"{'Layer Name':<40} | {'Max Abs Error':<15} | {'Rel L2 Error':<15} | {'> 1e-4':<8} | {'Status'}")
    print("-" * 120)
    print(f"{'lstm1':<40} | {max_d:.4e} | {max_rd:.4e} | {exceeds:<8} | {status}")
    print("="*120)

if __name__ == "__main__":
    run()

```

```
Input: shape=torch.Size([64, 10, 50])
Input - min: -4.590487, max: 4.220885, mean: 0.002711
Input - has NaN: False, has Inf: False
============================================================

>>> Running Mode A (cuda/f32/eager) ...
  ✓ Mode A (cuda/f32/eager) executed successfully
  Output - min: -0.581809, max: 0.548356, mean: -0.007435
  Output - has NaN: False, has Inf: False

>>> Running Switch Dtype (cuda/f16/eager) ...
  ✓ Switch Dtype (cuda/f16/eager) executed successfully
  Output - min: -0.582031, max: 0.548340, mean: -0.007435
  Output - has NaN: False, has Inf: False

========================================================================================================================
Comparison: Mode A (cuda/f32/eager) vs Switch Dtype (cuda/f16/eager)
========================================================================================================================
Layer Name                               | Max Abs Error   | Rel L2 Error    | > 1e-4   | Status
------------------------------------------------------------------------------------------------------------------------
lstm1                                    | 3.3201e-04 | 4.1319e-04 | Yes      | DIFF > 1e-4
========================================================================================================================

```

### Versions

PyTorch version: 2.10.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.10.19 | packaged by conda-forge | (main, Jan 26 2026, 23:45:08) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-90-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.6.20
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090

Nvidia driver version: 560.35.03
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 40 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 48
On-line CPU(s) list: 0-47
Vendor ID: GenuineIntel
Model name: QEMU Virtual CPU version 2.5+
CPU family: 15
Model: 107
Thread(s) per core: 1
Core(s) per socket: 48
Socket(s): 1
Stepping: 1
BogoMIPS: 4190.15
Flags: fpu de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx lm constant_tsc nopl xtopology cpuid tsc_known_freq pni ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c hypervisor lahf_lm abm cpuid_fault pti bmi1 avx2 bmi2 avx512f avx512dq avx512cd avx512bw avx512vl
Hypervisor vendor: KVM
Virtualization type: full
L1d cache: 1.5 MiB (48 instances)
L1i cache: 1.5 MiB (48 instances)
L2 cache: 192 MiB (48 instances)
L3 cache: 16 MiB (1 instance)
NUMA node(s): 1
NUMA node0 CPU(s): 0-47
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit: KVM: Mitigation: VMX unsupported
Vulnerability L1tf: Mitigation; PTE Inversion
Vulnerability Mds: Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown: Mitigation; PTI
Vulnerability Mmio stale data: Unknown: No mitigations
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Retpolines; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Retpoline
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected
Vulnerability Vmscape: Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] onnxruntime-gpu==1.23.2
[pip3] optree==0.18.0
[pip3] pytorch-triton==3.2.0+git4b3bb1f8
[pip3] torch==2.10.0+cu126
[pip3] torchaudio==2.11.0.dev20260127+cu126
[pip3] torchvision==0.25.0+cu126
[pip3] triton==3.6.0+git9844da95
[conda] numpy 1.26.4 pypi_0 pypi
[conda] nvidia-cublas-cu12 12.6.4.1 pypi_0 pypi
[conda] nvidia-cuda-cupti-cu12 12.6.80 pypi_0 pypi
[conda] nvidia-cuda-nvrtc-cu12 12.6.77 pypi_0 pypi
[conda] nvidia-cuda-runtime-cu12 12.6.77 pypi_0 pypi
[conda] nvidia-cudnn-cu12 9.10.2.21 pypi_0 pypi
[conda] nvidia-cufft-cu12 11.3.0.4 pypi_0 pypi
[conda] nvidia-curand-cu12 10.3.7.77 pypi_0 pypi
[conda] nvidia-cusolver-cu12 11.7.1.2 pypi_0 pypi
[conda] nvidia-cusparse-cu12 12.5.4.2 pypi_0 pypi
[conda] nvidia-cusparselt-cu12 0.7.1 pypi_0 pypi
[conda] nvidia-nccl-cu12 2.27.5 pypi_0 pypi
[conda] nvidia-nvjitlink-cu12 12.6.85 pypi_0 pypi
[conda] nvidia-nvtx-cu12 12.6.77 pypi_0 pypi
[conda] optree 0.18.0 pypi_0 pypi
[conda] pytorch-triton 3.2.0+git4b3bb1f8 pypi_0 pypi
[conda] torch 2.10.0+cu126 pypi_0 pypi
[conda] torchaudio 2.11.0.dev20260127+cu126 pypi_0 pypi
[conda] torchvision 0.25.0+cu126 pypi_0 pypi
[conda] triton 3.6.0+git9844da95 pypi_0 pypi

cc @mikaylagawarecki @ptrblck @msaroufim @eqy @jerryzh168 @tinglvv @nWEIdia
